### PR TITLE
[4.3] Quickicons

### PR DESCRIPTION
**Pull request by brianteeman**
This is an update of the abandoned PR #32671 with conflicts fixed and codestyle updated.

### Summary of Changes
See the original PR for further details and testing instructions

### CODE REVIEW

File: administrator/language/en-GB/mod_quickicon.ini [hunk 1]
```diff
@@ -6,18 +6,26 @@
...
+MOD_QUICKICON_CONTACT_MANAGER_ADD="Add Contact"
 MOD_QUICKICON_CONTENT="Content"
+MOD_QUICKICON_FEATURED_MANAGER="Featured Articles"
 MOD_QUICKICON_GLOBAL_CONFIGURATION="Global Configuration"
 MOD_QUICKICON_GROUP_DESC="The group of this module (this value is compared with the group value used in <strong>Quick Icons</strong> plugins to inject icons). The Joomla! core icons are available to be used in any Quick Icons module."
 MOD_QUICKICON_GROUP_LABEL="Group"
 MOD_QUICKICON_HEADER_ICON_LABEL="Header Icon"
 MOD_QUICKICON_INSTALL_EXTENSIONS="Install Extensions"
+MOD_QUICKICON_LANGUAGES_MANAGER="Languages"
+MOD_QUICKICON_LANGUAGES_MANAGER_ADD	="Add Language"
```
Comment (heelc29) [inline administrator/language/en-GB/mod_quickicon.ini]:
```suggestion
MOD_QUICKICON_LANGUAGES_MANAGER_ADD="Add Language"
```

File: administrator/language/en-GB/com_content.ini [hunk 2]
```diff
@@ -47,6 +47,10 @@ COM_CONTENT_ERROR_CANNOT_UNPUBLISH="One or more of the selected articles can't b
 COM_CONTENT_ERROR_UPDATE_STAGE="You cannot execute this transition and update the stage."
 COM_CONTENT_FEATURED_ARTICLES="Featured Articles"
 COM_CONTENT_FEATURED_CATEGORIES_LABEL="Select Categories"
+COM_CONTENT_FEATURED_N_QUICKICON="Featured Articles"
+COM_CONTENT_FEATURED_N_QUICKICON_1="Featured Article"
+COM_CONTENT_FEATURED_N_QUICKICON_SRONLY="Featured Articles: %s featured articles are published."
+COM_CONTENT_FEATURED_N_QUICKICON_SRONLY_1="Featured Article: One featured is published."
```
Comment (heelc29) [inline administrator/language/en-GB/com_content.ini]: Add `article` here?
```suggestion
COM_CONTENT_FEATURED_N_QUICKICON_SRONLY_1="Featured Article: One featured article is published."
```

File: administrator/components/com_contact/src/Controller/ContactsController.php [hunk 3]
```diff
@@ -13,6 +13,7 @@
 use Joomla\CMS\Application\CMSApplication;
 use Joomla\CMS\Language\Text;
 use Joomla\CMS\MVC\Controller\AdminController;
+use Joomla\CMS\Response\JsonResponse;
 use Joomla\CMS\MVC\Factory\MVCFactoryInterface;
```
Comment (Quy) [inline administrator/components/com_contact/src/Controller/ContactsController.php]:
```suggestion
use Joomla\CMS\MVC\Controller\AdminController;
use Joomla\CMS\MVC\Factory\MVCFactoryInterface;
use Joomla\CMS\Response\JsonResponse;
```

**Comment by Kostelano** [comment]
What is the principle of sorting in the module parameters and directly in the panel? Very chaotic.

Screenshot_2

Screenshot_1

File: administrator/language/en-GB/mod_quickicon.ini [hunk 4]
```diff
@@ -6,18 +6,26 @@
...
 MOD_QUICKICON_CACHE="Cache"
 MOD_QUICKICON_CATEGORY_MANAGER="Article Categories"
 MOD_QUICKICON_CATEGORY_MANAGER_ADD="Add Article Category"
 MOD_QUICKICON_CHECKIN="Checkin"
 MOD_QUICKICON_CHECKINS="Global Checkin"
 MOD_QUICKICON_CONFIGURATION="Configuration"
+MOD_QUICKICON_CONTACT_MANAGER="Contacts"
+MOD_QUICKICON_CONTACT_MANAGER_ADD="Add Contact"
 MOD_QUICKICON_CONTENT="Content"
+MOD_QUICKICON_FEATURED_MANAGER="Featured Articles"
```
Comment (Kostelano) [inline administrator/language/en-GB/mod_quickicon.ini:21]: Missed

```suggestion
MOD_QUICKICON_FEATURED_MANAGER="Featured Articles"
MOD_QUICKICON_FINDER_MANAGER="Smart Search"
```

File: administrator/language/en-GB/mod_quickicon.ini [hunk 5]
```diff
@@ -26,35 +34,50 @@ MOD_QUICKICON_MENUITEMS_MANAGER_ADD="Add Menu Item"
...
 MOD_QUICKICON_SHOW_USERS_LABEL="User Icon"
+MOD_QUICKICON_SHOW_WORKFLOW_LABEL="Workflows Icon"
 MOD_QUICKICON_SITE="Icons Site"
 MOD_QUICKICON_STRUCTURE="Structure"
 MOD_QUICKICON_SYSTEM="Icons System"
 MOD_QUICKICON_TAGS_MANAGER="Tags"
 MOD_QUICKICON_TAGS_MANAGER_ADD="Add Tag"
 MOD_QUICKICON_TEMPLATE_CODE="Template Code"
-MOD_QUICKICON_TEMPLATE_STYLES="Template Styles"
+MOD_QUICKICON_TEMPLATE_STYLES="Templates"
```
Comment (Kostelano) [inline administrator/language/en-GB/mod_quickicon.ini]: This is a link to styles. Probably the old line of text was correct.

**Comment by brianteeman** [comment]
> What is the principle of sorting in the module parameters and directly in the panel? Very chaotic.

Agreed. I couldnt see any logic in the original PR.

Happy to look at any suggestions - but not alphabetical

**Comment by chmst** [comment]
@Kostelano there is no logic.  And as every site is individual and need some buttons but does not need other buttons, finding a structure is  not as easy. I cannot imagine a site with all buttons activated. 
For example: 
I don't need users because I have no user registration. 
I have all plugins installed and used  the plugins button only during development (same for template buttons).
During development I don't need a cache or checkin function. 

So we cannot make a structure here.
Maybe "used in development only", or "buttons with count"?

I still think that the wohle feature needs complete re-writing with individual ordering and individual classes but meanwhile it is an improvment if there are more icons available.

**Comment by Kostelano** [comment]
I agree that "the only correct structure" is simply impossible to find, but there are some common things to understand - for example, that featured articles should be somewhere near regular articles, etc.

I played a little in photoshop. I don’t pretend to be the truth, but, let’s say, such a strauture (namely, the SEQUENCE of points, taking into account the possible disabling of some points or the FULL inclusion of all points, which is actually unlikely) would completely suit me. Although in the process of drawing, I moved points back and forth several times when I did not understand what I wanted to see.

**Current view:**

Screenshot_1

**My suggestion:**

Screenshot_1 — копия

Please, I ask visitors to take part in some kind of feedback, because in my opinion it is an interesting and necessary PR. Thank you.

**Comment by brianteeman** [comment]
> I still think that the wohle feature needs complete re-writing with individual ordering and individual classes but meanwhile it is an improvment if there are more icons available.

That is exactly my thought on this. Merge this now and then rewrite it with ordering later.

**Comment by Fedik** [comment]
It also can be a subform, with dropdown which "thing" to show. Then everyone can do their own ordering :)
But that need much changes in PR.

**Comment by chmst** [comment]
@Kostelano I like your ordering.

**Comment by Kostelano** [comment]
@brianteeman If you support my proposal to change the order of elements, I am ready to make a change in your branch. I just don't want to waste my time ;)

**Comment by brianteeman** [comment]
I think you will find its not as simple as you think it will be

**Comment by chmst** [comment]
If we can get two success tests, we can merge this for now and then you can make a PR for ordering @kostelano

**Comment by Kostelano** [comment]
I have tested this item :white_check_mark: successfully on 26f4aab819bf90f8d2e556006529e23af0c063d7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38267">issues.joomla.org/tracker/joomla-cms/38267</a>.</sub>

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on 26f4aab819bf90f8d2e556006529e23af0c063d7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38267">issues.joomla.org/tracker/joomla-cms/38267</a>.</sub>

**Comment by chmst** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38267">issues.joomla.org/tracker/joomla-cms/38267</a>.</sub>

**Comment by chmst** [comment]
@brianteeman there are unresolved conversations and block merging. Could you please check?

**Comment by brianteeman** [comment]
@chmst stupid github didnt close the conversations when the code was corrected

**Comment by fancyFranci** [comment]
It is a really nice PR, but I have to say that 4.3 is a better place for these improvements. I will change the base branch. Thank you for finishing it!

**Comment by brianteeman** [comment]
at the time the code was written and made rtc 4.2 was the correct place

**Comment by obuisard** [comment]
I like this PR, it may add clutter to the interface, but that is at the discretion of the administrator to do what it's right for the website managed. It is really nice to be able to have a full toolbox of all possible shortcuts. 

We will need to complement this PR with 2 other ones:
- allow for the ordering of the selected shortcuts,
- replace the old icon classes "icon-xxx" with "fa fa-xxx" in the quickicon module to match the new way icons are implemented in Joomla 4 - this is not something introduced in this PR, it is already something that needs to be fixed in the official release as well.

**Comment by brianteeman** [comment]
> allow for the ordering of the selected shortcuts,

good luck with that

**Comment by chmst** [comment]
@obuisard Ordering is a big task and not in scope of this PR. It would need a complete rework / another concept.

**Comment by obuisard** [comment]
> @obuisard Ordering is a big task and not in scope of this PR. It would need a complete rework / another concept.

Indeed, that is why we need to create 2 **other** PRs to improve on this one :-)
And I agree, the ordering of the quickicons is challenging

**Comment by brianteeman** [comment]
until this is merged no one can even try

**Comment by N6REJ** [comment]
> * replace the old icon classes "icon-xxx" with "fa fa-xxx" in the quickicon module to match the new way icons are implemented in Joomla 4 - this is not something introduced in this PR, it is already something that needs to be fixed in the official release as well.

That was done last year or so but this pr seems to be undoing it.

**Comment by obuisard** [comment]
> until this is merged no one can even try

I won't blindly merge PRs.

**Comment by obuisard** [comment]
> > * replace the old icon classes "icon-xxx" with "fa fa-xxx" in the quickicon module to match the new way icons are implemented in Joomla 4 - this is not something introduced in this PR, it is already something that needs to be fixed in the official release as well.
> 
> That was done last year or so but this pr seems to be undoing it.

Is it @N6REJ ? because I don't see the fa fa-xxx icons on my 4.2 version

**Comment by brianteeman** [comment]
> > until this is merged no one can even try
> 
> I won't blindly merge PRs.

would object if you did

**Comment by N6REJ** [comment]
> > > * replace the old icon classes "icon-xxx" with "fa fa-xxx" in the quickicon module to match the new way icons are implemented in Joomla 4 - this is not something introduced in this PR, it is already something that needs to be fixed in the official release as well.
> > 
> > 
> > That was done last year or so but this pr seems to be undoing it.
> 
> Is it @N6REJ ? because I don't see the fa fa-xxx icons on my 4.2 version

There were 3 large pr's that added fa- to every J! icon.  At that time they were the "We're using fontawesome from now on" icon set.  but we were told for b/c we had to also provide ability for standard icon-
Since then everything has been moved to a js build so I've no clue where everything went nor what happened.

**Comment by obuisard** [comment]
> > * replace the old icon classes "icon-xxx" with "fa fa-xxx" in the quickicon module to match the new way icons are implemented in Joomla 4 - this is not something introduced in this PR, it is already something that needs to be fixed in the official release as well.
> 
> That was done last year or so but this pr seems to be undoing it.

For reference, the change to -icons was made October 2020 to remove the dependency on FontAwesome. #30707

**Comment by brianteeman** [comment]
this pr is already using icon- notation so I dot even know why it is being discussed.

**Comment by brianteeman** [comment]
> > > > * replace the old icon classes "icon-xxx" with "fa fa-xxx" in the quickicon module to match the new way icons are implemented in Joomla 4 - this is not something introduced in this PR, it is already something that needs to be fixed in the official release as well.
> > > 
> > > 
> > > That was done last year or so but this pr seems to be undoing it.
> > 
> > 
> > Is it @N6REJ ? because I don't see the fa fa-xxx icons on my 4.2 version
> 
> There were 3 large pr's that added fa- to every J! icon. At that time they were the "We're using fontawesome from now on" icon set. but we were told for b/c we had to also provide ability for standard icon- Since then everything has been moved to a js build so I've no clue where everything went nor what happened.

I think you are confused. The PRs you created were not merged https://github.com/joomla/joomla-cms/pulls?q=is%3Apr+author%3AN6REJ+is%3Aclosed+icon+is%3Aunmerged+

**Comment by N6REJ** [comment]
lets start with a simple one https://github.com/joomla/joomla-cms/pull/27686

**Comment by N6REJ** [comment]
@brianteeman are you introducing any icons that don't pre-exist in the fa- compatibility table?

**Comment by brianteeman** [comment]
Read the code and test the pr and stop wasting everyone's time. This bike shedding is why it takes so long to move forward.

**Comment by N6REJ** [comment]
> > > * replace the old icon classes "icon-xxx" with "fa fa-xxx" in the quickicon module to match the new way icons are implemented in Joomla 4 - this is not something introduced in this PR, it is already something that needs to be fixed in the official release as well.
> > 
> > 
> > That was done last year or so but this pr seems to be undoing it.
> 
> Is it @N6REJ ? because I don't see the fa fa-xxx icons on my 4.2 version

just an fyi, per our conversation you can still use fa- instead of icon- in admin if you desire.
image

**Comment by obuisard** [comment]
Thank you very much Brian @brianteeman for the PR and all who participated in the conversation. Thank you @Bongovo for the original PR.

**Pull request merged**